presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java

```diff
@@ -193,18 +200,41 @@ public Page getOutput()
         }
 
         if (nestedLoopPageBuilder != null && nestedLoopPageBuilder.hasNext()) {
-            return nestedLoopPageBuilder.next();
+            return applyFilter(nestedLoopPageBuilder.next());
         }
 
         if (buildPageIterator.hasNext()) {
             nestedLoopPageBuilder = new NestedLoopPageBuilder(probePage, buildPageIterator.next());
-            return nestedLoopPageBuilder.next();
+            return applyFilter(nestedLoopPageBuilder.next());
         }
 
         probePage = null;
         return null;
     }
 
+    private Page applyFilter(Page page)
+    {
+        if (!pageFilter.isPresent()) {
+            return page;
+        }
+
+        SelectedPositions selectedPositions = pageFilter.get().filter(operatorContext.getSession().toConnectorSession(), pageFilter.get().getInputChannels().getInputChannels(page));
+        Block[] blocks = page.getBlocks();
+        Block[] filteredBlocks = new Block[blocks.length];
+        if (selectedPositions.isList()) {
+            for (int i = 0; i < blocks.length; i++) {
+                filteredBlocks[i] = blocks[i].copyPositions(selectedPositions.getPositions(), selectedPositions.getOffset(), selectedPositions.size());
+            }
+        }
+        else {
+            for (int i = 0; i < blocks.length; i++) {
```

Comment:
Why not simply return:
```
return page.getRegion(selectedPositions.getOffset(), selectedPositions.size())
```
`Page` has more information than `Block` so it could potentially do `getRegion` more efficiently in the future.

Reply:
@findepi, @sopel39, I didn't see getRegion and getPositions methods in Page class. Thanks for pointing them out. Will update.